storage/entry_cache.go

```diff
@@ -0,0 +1,125 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Kevin GuanJian (keynovo@gmail.com)
+
+package storage
+
+import (
+	"sync"
+
+	"github.com/cockroachdb/cockroach/util/cache"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+// A raftEntryCache maintains the recent entries
+type raftEntryCache struct {
+	// entryCache caches recent entries, the Index as the key.
+	entryCache *cache.UnorderedCache
+	// the minimum Index in the cache, evicted when cache is filled.
+	firstIndex uint64
+	// the maximum size of the cache.
+	size int
+	// protects Cache for concurrent access.
+	sync.Mutex
+}
+
+// newRaftEntryCache returns a new RaftEntryCache with the given size.
+func newRaftEntryCache(size int) *raftEntryCache {
+	return &raftEntryCache{
+		entryCache: cache.NewUnorderedCache(cache.Config{
+			Policy: cache.CacheNone,
+		}),
+		size: size,
+	}
+}
+
+// addEntry will add entries, the Index as the key.
+func (rec *raftEntryCache) addEntry(ents []raftpb.Entry) {
+	rec.Lock()
+	defer rec.Unlock()
+	if len(ents) == 0 {
+		return
+	}
+
+	for _, e := range ents {
+		rec.entryCache.Add(e.Index, e)
+	}
+
+	if rec.firstIndex == uint64(0) ||
+		rec.firstIndex > ents[0].Index {
+		rec.firstIndex = ents[0].Index
+	}
+	rec.evictEntry()
+}
+
+// getEntry will return entries between [lo, hi).
+func (rec *raftEntryCache) getEntry(lo, hi, maxBytes uint64) ([]raftpb.Entry, uint64 /* size */, uint64 /* max Index */) {
+	rec.Lock()
+	defer rec.Unlock()
+	var entries []raftpb.Entry
+	var ent raftpb.Entry
+	size := uint64(0)
+
+	for i := lo; i < hi; i++ {
+		v, ok := rec.entryCache.Get(i)
+		if ok {
+			ent = v.(raftpb.Entry)
+			entries = append(entries, ent)
+			size = size + uint64(ent.Size())
+			if maxBytes > 0 && size > maxBytes {
+				return entries, size, i + 1
+			}
+		} else {
+			return entries, size, i
+		}
+	}
+	return entries, size, hi
+}
+
+// delEntry will delete any entries between [lo, hi).
+func (rec *raftEntryCache) delEntry(lo, hi uint64) {
+	rec.Lock()
+	defer rec.Unlock()
+	if lo > hi {
+		return
+	}
+	for i := lo; i < hi; i++ {
+		rec.entryCache.Del(i)
+	}
+}
+
+// evictEntry will evict the entries from the minimum Index
+// when the cache filled.
+func (rec *raftEntryCache) evictEntry() {
+	if len := rec.entryCache.Len(); len <= rec.size {
+		return
+	} else if len > rec.size {
+		evictLen := len - rec.size
+		for i := rec.firstIndex; i < rec.firstIndex+uint64(evictLen); i++ {
```

Comment:
Ah yes, you'd still need a `firstIndex` set on truncation, but you don't need to purge old entries from the cache when it is updated; you can just filter them out on read and wait for the regular LRU eviction.

However, if you don't clear the cache immediately on truncation, then the cache will stay at its maximum size for each range, which is expensive. I think we can only rely on LRU eviction if we move from a per-range cache to a global one. I still think that a global cache is a good idea because of the bounded worst-case memory consumption.